### PR TITLE
[doc] Add API stability annotations to developer guide

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds a contributor guide on how to annotate public APIs and style suggestions on improving compatibility.

### CODE REVIEW

File: doc/source/getting-involved.rst [hunk 1]
```diff
@@ -219,6 +219,45 @@ check recent tests known to be flaky.
...
+* ``@PublicAPI(stability="beta")``
+    * These APIs are in beta but will eventually be stabilized.
+
+* ``@DeveloperAPI``
+    * Low-level integration APIs that may change across minor Ray versions.
+
+* ``@Deprecated``
+    * This API is deprecated and may be removed in future Ray versions.
+
+Undecorated functions can be generally assumed to not be part of the Ray public API.
```
Comment (edoakes) [inline doc/source/getting-involved.rst:239]: should these be pulled from the docstrings of the decorators?

**Review by edoakes — APPROVED** [review]

File: doc/source/getting-involved.rst (hunk 1 above)
Comment (ericl) [reply doc/source/getting-involved.rst:239]: Let me try.

File: doc/source/getting-involved.rst [hunk 2]
```diff
@@ -219,6 +219,37 @@ check recent tests known to be flaky.
...
 .. _`CI`: https://github.com/ray-project/ray/tree/master/ci/travis
 
+API stability
+-------------
+
+Ray provides stability guarantees for its public APIs in Ray core and libraries. The level of stability provided depends on how the API is annotated.
+
+.. autofunction:: ray.util.annotations.PublicAPI
+.. autofunction:: ray.util.annotations.DeveloperAPI
+.. autofunction:: ray.util.annotations.Deprecated
```
Comment (edoakes) [inline doc/source/getting-involved.rst:229]: Nice! Looks pretty good in the generated docs.

Comment (richardliaw) [reply doc/source/getting-involved.rst:229]: Maybe in the "Deprecated" section we can talk about what the expectation would be for deprecation?

For example, can you expect the deprecation to happen over the course of 1 release, 2 releases? Is it always expected to be the case that the annotation will be applied before removal? etc

**Review by richardliaw — APPROVED** [review]
Looks good, 1 suggestion

**Pull request merged**